# The Battle of the Neighborhoods
### Applied Data Science Capstone Coursera

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Results](#results)
* [Discussion](#discussion)
* [Conclusion](#conclusion)

## Introduction: Business Problem <a name="introduction"></a>

This notebook will try to find similar neighborhoods in two different cities based on the type of venues and popular spots contained in the neighborhoods. Specifically, we will look at neighborhoods in two major cities; (Manhattan) New York, New York USA and Toronto, Ontario Canada.

This project would be of interest to people moving from Toronto proper to Manhattan and are looking to settle in similar areas. This project would also be of interest to companies looking to move headquarters or open up branch offices in Manhattan in areas that are similar to the ones they currently operate in in Toronto.

## Data <a name="data"></a>

In order to complete our project we will need the following data:

* Coordinate Data for the city centers of Manhattan and Toronto, available from **GeoPy Nominatim**
* Coordinate Data for the neighborhoods in Manhattan, available at this URL: **https://cocl.us/new_york_dataset**
* Neighborhoods in Toronto by postal code, available on **Wikipedia: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M**
* Coordinate Data for postal codes in Toronto, available at this URL: **http://cocl.us/Geospatial_data**
* Coordinate Data and Venue/spot type from **Foursquare API**

Using the Foursquare API data we will classify the various neighborhoods of Manhattan based on the most common types of venues in that neighborhood.
Then, we will do the same with the various neighborhoods of Toronto proper. Finally, we will compare the neighborhoods to determine which are 'closest' to one another by examining a matrix of the relative frequencies of each type of venue.

### Libraries
Let's start by importing all the various libraries necessary to complete our project.

In [2]:
import numpy as np 

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 

import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans

import folium
!conda install lxml -y

from sklearn import metrics

from sklearn.neighbors import KNeighborsClassifier

!pip install yellowbrick
from yellowbrick.cluster import KElbowVisualizer

print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         238 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0        conda-forge
    geopy:         1.20.0-py_0      conda-forge

The following pac

### Manhattan Data
In the following code will download the Manhattan neighborhood data and save this into a Pandas dataframe.
Then we will use **Nominatim** and **Folium** to generate a map of Manhattan with the neighborhoods' locations marked.
Finally, we will make calls to the **Foursquare API** to generate a table containing venue data.

In [ ]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

In [3]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [4]:
neighborhoods_data = newyork_data['features']

In [5]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
nyc = pd.DataFrame(columns=column_names)

for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    nyc = nyc.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [93]:
nyc.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [94]:
manhattan = nyc[nyc['Borough'] == 'Manhattan'].reset_index(drop = True)

In [95]:
manhattan.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

In [96]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Manhattan are 40.7896239, -73.9598939.


In [97]:
# create map of NYC using latitude and longitude values
map_nyc = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan['Latitude'], manhattan['Longitude'], manhattan['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_nyc)  
    
map_nyc

In [137]:
CLIENT_ID = 'L'
CLIENT_SECRET = 'R'
VERSION = '20180605' 
radius = 500
LIMIT = 100

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: L
CLIENT_SECRET:R


In [12]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [13]:
manhattan_venues = getNearbyVenues(names=manhattan['Neighborhood'],
                                   latitudes=manhattan['Latitude'],
                                   longitudes=manhattan['Longitude']
                                  )


Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


In [14]:
manhattan_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066      Starbucks   
4  Marble Hill              40.876551               -73.91066        Dunkin'   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877531       -73.905582    Coffee Shop  
4       40.877136       -73.906666     Donut Shop

In [15]:
print('There are {} unique categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 340 unique categories.


### Toronto Data
In the following code will download the Toronto neighborhood data from **Wikipedia** and save this into a Pandas dataframe along with coordinate data from this URL **http://cocl.us/Geospatial_data**.
Then we will use **Nominatim** and **Folium** to generate a map of Toronto with the neighborhoods' locations marked.
Finally, we will make calls to the **Foursquare API** to generate a table containing venue data.

In [16]:
# read the table off the webpage
link = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
tables = pd.read_html(link)

In [17]:
# first item in the list is the table needed
type(tables)
df = tables[0]
df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [18]:
# drop entries with an unassigned borough
df = df[df['Borough']!= 'Not assigned'].reset_index(drop = True)

In [19]:
# assign borough name to neighborhood name for single entry
df[df['Neighbourhood'] == 'Not assigned']
df.iloc[6][2] = df.iloc[6][1]

In [20]:
# clean up column names
df.rename(columns = {'Postcode':'PostalCode', 'Neighbourhood':'Neighborhood'}, inplace = True)
df.sort_values('PostalCode', inplace = True, ascending = True)
df = df.reset_index(drop = True)

In [21]:
toronto = df[df['Borough'].str.contains("Toronto")].reset_index(drop = True)

In [22]:
# grouping by postcode and borough put neighborhoods together
toronto = toronto.groupby(['PostalCode','Borough'], as_index = False, sort = False).agg(", ".join)

In [23]:
# download coordinates
!wget -q -O 'latlon.csv' http://cocl.us/Geospatial_data

latlon_df = pd.read_csv('latlon.csv')

latlon_df.rename(columns = {'Postal Code': 'PostalCode'}, inplace = True)

# merge dataframes to obtain final dataframe
toronto = pd.merge(toronto, latlon_df, how = 'inner', on = 'PostalCode')


In [24]:
toronto.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [25]:
# use geolocator to get coordinates for Toronto
t_address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="to_explorer")
t_location = geolocator.geocode(t_address)
t_latitude = t_location.latitude
t_longitude = t_location.longitude
print('The geograpical coordinates of Toronto, ON are {}, {}.'.format(t_latitude, t_longitude))

The geograpical coordinates of Toronto, ON are 43.655115, -79.380219.


In [26]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[t_latitude, t_longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough'], toronto['Neighborhood']):
    label = '{}; {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [27]:
toronto_venues = getNearbyVenues(names = toronto['Neighborhood'],
                                   latitudes = toronto['Latitude'],
                                   longitudes = toronto['Longitude']
                                  )


The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
South Hill, Rathnelly, Summerhill West, Deer Park, Forest Hill SE
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Regent Park, Harbourfront
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
King, Adelaide, Richmond
Toronto Islands, Harbourfront East, Union Station
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, Yorkville, North Midtown
University of Toronto, Harbord
Kensington Market, Grange Park, Chinatown
Railway Lands, South Niagara, Harbourfront West, Island airport, Bathurst Quay, CN Tower, King and Spadina
Stn A PO Boxes 25 The Esplanade
Underground city, First Canadian Place
Christie
Dovercourt Village, Dufferin
Trinity, Little Portugal
Exhibition Place, Brockton, Parkdale Village
High Park, The 

In [28]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

In [29]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 233 unique categories.


We can see that there are more unique venue categories in Manhattan than in Toronto so we will have to drop any venues with venue categories not shared by both cities.

In [30]:
manhattan_venues_shared = manhattan_venues[manhattan_venues['Venue Category'].isin(toronto_venues['Venue Category'].unique())].reset_index(drop = True)

In [31]:
toronto_venues_shared = toronto_venues[toronto_venues['Venue Category'].isin(manhattan_venues['Venue Category'].unique())].reset_index(drop = True)

In [32]:
print('There are {} unique categories.'.format(len(toronto_venues_shared['Venue Category'].unique())))

There are 187 unique categories.


In [33]:
print('There are {} unique categories.'.format(len(manhattan_venues_shared['Venue Category'].unique())))

There are 187 unique categories.


## Methodology <a name="methodology"></a>

Now that we have our dataframes containing venue data from each city let's use one hot encoding to count the number of venues of each venue category in each neighborhood. 

In [34]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues_shared[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot = pd.get_dummies(toronto_venues_shared[['Venue Category']], prefix="", prefix_sep="")

In [35]:
# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues_shared['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Afghan Restaurant  American Restaurant  Antique Shop  \
0  Marble Hill                  0                    0             0   
1  Marble Hill                  0                    0             0   
2  Marble Hill                  0                    0             0   
3  Marble Hill                  0                    0             0   
4  Marble Hill                  0                    0             0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  BBQ Joint  \
0            0           0                    0                 0          0   
1            0           0                    0                 0          0   
2            0           0                    0                 0          0   
3            0           0                    0                 0          0   
4            0           0                    0                 0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Bed & Breakfast  Beer Bar  \
0           0           0       0     0    0                0         0   
1           0           0       0     0    0                0         0   
2           0           0       0     0    0                0         0   
3           0           0       0     0    0                0         0   
4           0           0       0     0    0                0         0   

   Beer Store  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0       0              0          0         0   
1           0       0              0          0         0   
2           0       0              0          0         0   
3           0       0              0          0         0   
4           0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Bubble Tea Shop  Building  \
0                     0               0                0         0   
1                     0               0                0         0   
2                     0               0                0         0   
3                     0               0                0         0   
4                     0               0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0                          0             0                     0            0   
1                          0             0                     0            0   
2                          0             0                     0            0   
3                          0             0                     0            0   
4                          0             0                     0            0   

   Chinese Restaurant  Chocolate Shop  Climbing Gym  Clothing Store  \
0                   0               0             0               0   
1                   0               0             0               0   
2                   0               0             0               0   
3                   0               0             0               0   
4                   0               0             0               0   

   Cocktail Bar  Coffee Shop  Concert Hall  Convenience Store  Cosmetics Shop  \
0             0            0             0                  0               0   
1             0            0             0                  0               0   
2             0            0             0                  0               0   
3             0            1             0                  0               0   
4             0            0             0                  0               0   

   Coworking Space  Creperie  Cuban Restaurant  Cupcake Shop  Dance Studio  \
0              

In [36]:
manhattan_onehot.shape

(2916, 188)

In [37]:
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues_shared['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Neighborhood  Afghan Restaurant  American Restaurant  \
0                   The Beaches                  0                    0   
1                   The Beaches                  0                    0   
2                   The Beaches                  0                    0   
3  The Danforth West, Riverdale                  0                    0   
4  The Danforth West, Riverdale                  0                    0   

   Antique Shop  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0            0           0                    0   
1             0            0           0                    0   
2             0            0           0                    0   
3             0            0           0                    0   
4             0            0           0                    0   

   Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                 0          0           0           0       0     0    0   
1                 0          0           0           0       0     0    0   
2                 0          0           0           0       0     0    0   
3                 0          0           0           0       0     0    0   
4                 0          0           0           0       0     0    0   

   Bed & Breakfast  Beer Bar  Beer Store  Bistro  Boat or Ferry  Bookstore  \
0                0         0           0       0              0          0   
1                0         0           0       0              0          0   
2                0         0           0       0              0          0   
3                0         0           0       0              0          0   
4                0         0           0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Bubble Tea Shop  Building  \
0         0                     0               0                0         0   
1         0                     0               0                0         0   
2         0                     0               0                0         0   
3         0                     0               0                0         0   
4         0                     0               0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0                          0             0                     0            0   
1                          0             0                     0            0   
2                          0             0                     0            0   
3                          0             0                     0            0   
4                          0             0                     0            0   

   Chinese Restaurant  Chocolate Shop  Climbing Gym  Clothing Store  \
0                   0               0             0               0   
1                   0               0             0               0   
2                   0               0             0               0   
3                   0               0             0               0   
4                   0               0             0               0   

   Cocktail Bar  Coffee Shop  Concert Hall  Convenience Store  Cosmetics Shop  \
0             0            0             0                  0               0   
1             0            0             0                  0               0   
2             0            0             0                  0               0   
3             0            0             0                  0               0   
4             0            0             0                  0               0   

   Cowor

In [38]:
toronto_onehot.shape

(1607, 188)

Now, we will calculate the mean of each venue category to the total number of venues in each neighborhood. This will tell us the relative frequency of that venue category to the total venues in that neighborhood. A value of 1.0 would represent that all venues in that neighborhood are of that category. A value of 0 would indicate that no venues of that category exist in that neighborhood. 
Let's define how similar two neighborhoods are by how closely their column values match each other.

In [39]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Afghan Restaurant  American Restaurant  Antique Shop  \
0     Battery Park City           0.000000             0.011905      0.000000   
1         Carnegie Hill           0.000000             0.011111      0.000000   
2        Central Harlem           0.000000             0.054054      0.000000   
3               Chelsea           0.000000             0.034091      0.000000   
4             Chinatown           0.000000             0.043478      0.000000   
5          Civic Center           0.000000             0.032609      0.010870   
6               Clinton           0.000000             0.047059      0.000000   
7           East Harlem           0.000000             0.000000      0.000000   
8          East Village           0.000000             0.022989      0.011494   
9    Financial District           0.000000             0.032967      0.000000   
10             Flatiron           0.000000             0.044944      0.000000   
11             Gramercy           0.000000             0.045977      0.000000   
12    Greenwich Village           0.000000             0.021053      0.000000   
13     Hamilton Heights           0.000000             0.000000      0.000000   
14         Hudson Yards           0.000000             0.078125      0.000000   
15               Inwood           0.000000             0.042553      0.000000   
16           Lenox Hill           0.011765             0.000000      0.000000   
17       Lincoln Square           0.000000             0.022727      0.000000   
18         Little Italy           0.000000             0.011905      0.000000   
19      Lower East Side           0.000000             0.000000      0.000000   
20     Manhattan Valley           0.000000             0.000000      0.000000   
21       Manhattanville           0.000000             0.030303      0.000000   
22          Marble Hill           0.000000             0.045455      0.000000   
23              Midtown           0.000000             0.022472      0.000000   
24        Midtown South           0.000000             0.031915      0.000000   
25  Morningside Heights           0.000000             0.075000      0.000000   
26          Murray Hill           0.000000             0.031915      0.000000   
27                 Noho           0.000000             0.024390      0.000000   
28     Roosevelt Island           0.000000             0.000000      0.000000   
29                 Soho           0.000000             0.011494      0.000000   
30      Stuyvesant Town           0.000000             0.000000      0.000000   
31         Sutton Place           0.000000             0.034483      0.000000   
32              Tribeca           0.000000             0.059524      0.000000   
33           Tudor City           0.000000             0.013699      0.000000   
34           Turtle Bay           0.000000             0.011905      0.000000   
35      Upper East Side           0.000000             0.023256      0.000000   
36      Upper West Side           0.000000             0.022222      0.000000   
37   Washington Heights           0.000000             0.013889      0.000000   
38         West Village           0.000000             0.044444      0.000000   
39            Yorkville           0.000000             0.000000      0.000000   

    Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  BBQ Joint  \
0      0.000000    0.000000             0.000000          0.000000   0.023810   
1      0.000000    0.011111             0.000000          0.000000   0.000000   
2      0.027027    0.000000             0.000000          0.000000   0.027027   
3      0.022727    0.000000             0.000000          0.011364   0.000000   
4      0.000000    0.000000             0.000000          0.021739   0.000000   
5      0.021739    0.000000             0.000000          0.010870   0.000000   
6      0.000000    0.000000             0.000000          0.000000   0.000000   
7      0.000000    0.000000             0

In [40]:
# calculate mean number of each type of venue
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1   Business Reply Mail Processing Centre 969 Eastern           0.000000   
2                         Cabbagetown, St. James Town           0.000000   
3                                  Central Bay Street           0.000000   
4                                            Christie           0.000000   
5                                Church and Wellesley           0.011111   
6                      Commerce Court, Victoria Hotel           0.000000   
7                                          Davisville           0.000000   
8                                    Davisville North           0.000000   
9                        Dovercourt Village, Dufferin           0.000000   
10       Exhibition Place, Brockton, Parkdale Village           0.000000   
11                Forest Hill North, Forest Hill West           0.000000   
12                      High Park, The Junction South           0.000000   
13          Kensington Market, Grange Park, Chinatown           0.000000   
14                           King, Adelaide, Richmond           0.000000   
15                                      Lawrence Park           0.000000   
16                        Moore Park, Summerhill East           0.000000   
17                                 North Toronto West           0.000000   
18  Railway Lands, South Niagara, Harbourfront Wes...           0.000000   
19                          Regent Park, Harbourfront           0.000000   
20                             Roncesvalles, Parkdale           0.000000   
21                                           Rosedale           0.000000   
22                                           Roselawn           0.000000   
23                                 Runnymede, Swansea           0.000000   
24                           Ryerson, Garden District           0.000000   
25  South Hill, Rathnelly, Summerhill West, Deer P...           0.000000   
26                                     St. James Town           0.000000   
27                    Stn A PO Boxes 25 The Esplanade           0.000000   
28                                    Studio District           0.000000   
29                The Annex, Yorkville, North Midtown           0.000000   
30                                        The Beaches           0.000000   
31                     The Beaches West, India Bazaar           0.000000   
32                       The Danforth West, Riverdale           0.000000   
33           Toronto Dominion Centre, Design Exchange           0.000000   
34  Toronto Islands, Harbourfront East, Union Station           0.000000   
35                           Trinity, Little Portugal           0.000000   
36             Underground city, First Canadian Place           0.000000   
37                     University of Toronto, Harbord           0.000000   

    American Restaurant  Antique Shop  Art Gallery  Art Museum  \
0              0.000000      0.000000     0.019608    0.000000   
1              0.000000      0.000000     0.000000    0.000000   
2              0.000000      0.000000     0.000000    0.000000   
3              0.012048      0.000000     0.000000    0.012048   
4              0.000000      0.000000     0.000000    0.000000   
5              0.011111      0.000000     0.000000    0.000000   
6              0.040816      0.000000     0.010204    0.000000   
7              0.000000      0.000000     0.000000    0.000000   
8              0.000000      0.000000     0.000000    0.000000   
9              0.000000      0.000000     0.066667    0.000000   
10             0.000000      0.000000     0.000000    0.000000   
11             0.000000      0.000000     0.000000    0.000000   
12             0.000000      0.000000     0.000000    0.000000   
13             0.000000      0.000000     0.000000    0.000000   
14             0.031250      0.000000     0.010417    0.010417   
15             0.000

Now let's take a look at the 10 most common venue types in each neighborhood.

In [41]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
manhattan_sorted = pd.DataFrame(columns=columns)
manhattan_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    manhattan_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

manhattan_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park           Coffee Shop   
1      Carnegie Hill           Coffee Shop           Pizza Place   
2     Central Harlem    Chinese Restaurant        Cosmetics Shop   
3            Chelsea           Coffee Shop                Bakery   
4          Chinatown    Chinese Restaurant          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                 Hotel                   Gym     Italian Restaurant   
1        Cosmetics Shop                  Café            Yoga Studio   
2   American Restaurant  Gym / Fitness Center     Seafood Restaurant   
3    Italian Restaurant        Ice Cream Shop              Nightclub   
4                Bakery   American Restaurant  Vietnamese Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Food Court          Cupcake Shop        Clothing Store   
1                Bakery         Grocery Store     French Restaurant   
2                   Bar     French Restaurant              Boutique   
3                 Hotel   American Restaurant               Theater   
4    Salon / Barbershop          Optical Shop                   Spa   

  9th Most Common Venue 10th Most Common Venue  
0           Pizza Place             Playground  
1   Japanese Restaurant              Bookstore  
2           Event Space                    Spa  
3    Seafood Restaurant            Men's Store  
4       Bubble Tea Shop      Korean Restaurant

In [43]:
# create a new dataframe
toronto_sorted = pd.DataFrame(columns=columns)
toronto_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    toronto_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

toronto_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
2                        Cabbagetown, St. James Town           Coffee Shop   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie                  Café   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop    Italian Restaurant   
1         Burrito Place         Garden Center                Garden   
2                  Park           Pizza Place                  Café   
3                  Café    Italian Restaurant          Burger Joint   
4         Grocery Store                  Park            Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Farmers Market              Beer Bar            Steakhouse   
1       Fast Food Restaurant        Farmers Market                  Park   
2                     Market                   Pub    Italian Restaurant   
3  Middle Eastern Restaurant        Sandwich Place        Ice Cream Shop   
4                      Diner             Nightclub           Coffee Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant                  Café                 Bakery  
1            Restaurant            Skate Park             Smoke Shop  
2                Bakery            Restaurant          Grocery Store  
3    Chinese Restaurant                   Bar       Sushi Restaurant  
4            Baby Store     Convenience Store     Italian Restaurant

In [44]:
manhattan_merged = manhattan

# merge manhattan_grouped with manhattan to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(manhattan_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged

Borough         Neighborhood   Latitude  Longitude  \
0   Manhattan          Marble Hill  40.876551 -73.910660   
1   Manhattan            Chinatown  40.715618 -73.994279   
2   Manhattan   Washington Heights  40.851903 -73.936900   
3   Manhattan               Inwood  40.867684 -73.921210   
4   Manhattan     Hamilton Heights  40.823604 -73.949688   
5   Manhattan       Manhattanville  40.816934 -73.957385   
6   Manhattan       Central Harlem  40.815976 -73.943211   
7   Manhattan          East Harlem  40.792249 -73.944182   
8   Manhattan      Upper East Side  40.775639 -73.960508   
9   Manhattan            Yorkville  40.775930 -73.947118   
10  Manhattan           Lenox Hill  40.768113 -73.958860   
11  Manhattan     Roosevelt Island  40.762160 -73.949168   
12  Manhattan      Upper West Side  40.787658 -73.977059   
13  Manhattan       Lincoln Square  40.773529 -73.985338   
14  Manhattan              Clinton  40.759101 -73.996119   
15  Manhattan              Midtown  40.754691 -73.981669   
16  Manhattan          Murray Hill  40.748303 -73.978332   
17  Manhattan              Chelsea  40.744035 -74.003116   
18  Manhattan    Greenwich Village  40.726933 -73.999914   
19  Manhattan         East Village  40.727847 -73.982226   
20  Manhattan      Lower East Side  40.717807 -73.980890   
21  Manhattan              Tribeca  40.721522 -74.010683   
22  Manhattan         Little Italy  40.719324 -73.997305   
23  Manhattan                 Soho  40.722184 -74.000657   
24  Manhattan         West Village  40.734434 -74.006180   
25  Manhattan     Manhattan Valley  40.797307 -73.964286   
26  Manhattan  Morningside Heights  40.808000 -73.963896   
27  Manhattan             Gramercy  40.737210 -73.981376   
28  Manhattan    Battery Park City  40.711932 -74.016869   
29  Manhattan   Financial District  40.707107 -74.010665   
30  Manhattan        Carnegie Hill  40.782683 -73.953256   
31  Manhattan                 Noho  40.723259 -73.988434   
32  Manhattan         Civic Center  40.715229 -74.005415   
33  Manhattan        Midtown South  40.748510 -73.988713   
34  Manhattan         Sutton Place  40.760280 -73.963556   
35  Manhattan           Turtle Bay  40.752042 -73.967708   
36  Manhattan           Tudor City  40.746917 -73.971219   
37  Manhattan      Stuyvesant Town  40.731000 -73.974052   
38  Manhattan             Flatiron  40.739673 -73.990947   
39  Manhattan         Hudson Yards  40.756658 -74.000111   

   1st Most Common Venue    2nd Most Common Venue 3rd Most Common Venue  \
0            Coffee Shop           Sandwich Place           Yoga Studio   
1     Chinese Restaurant             Cocktail Bar                Bakery   
2                   Café            Grocery Store                Bakery   
3     Mexican Restaurant                     Café                Lounge   
4            Pizza Place                     Café    Mexican Restaurant   
5            Coffee Shop       Mexican Restaurant    Italian Restaurant   
6     Chinese Restaurant           Cosmetics Shop   American Restaurant   
7     Mexican Restaurant                   Bakery         Deli / Bodega   
8     Italian Restaurant                   Bakery           Art Gallery   
9            Coffee Shop                      Gym    Italian Restaurant   
10           Coffee Shop       Italian Restaurant           Pizza Place   
11                  Park              Coffee Shop        Sandwich Place   
12    Italian Restaurant              Coffee Shop              Wine Bar   
13  Gym / Fitness Center                  Theater                 Plaza   
14               Theater     Gym / Fitness Center   American Restaurant   
15                 Hotel           Clothing Store               Theater   
16           Coffee Shop                    Hotel        Sandwich Place   
17           Coffee Shop                   Bakery    Italian Restaurant   
18    Italian Restaurant           Clothing Store      Sushi Restaurant   
19                   Bar                 Wine

In [45]:
manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

In [46]:
X = manhattan_grouped_clustering.values
X.shape

(40, 187)

In [47]:
y = manhattan_grouped_clustering.index.values
y

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39])

In [49]:
Z = toronto_grouped_clustering.values
Z.shape

Since we are only concerned about the 'closest' neighborhood we only need to set k = 1 in the k-Nearest Neighbors algorithm.
Also, since the dimension of each element of our matrix is 187, we will simplify the calculation by only calculating the absolute difference of each column value to find the most similar Manhattan neighborhood to our Toronto neighborhoods.

In [50]:
k = 1
#Train Model and Predict  
neigh = KNeighborsClassifier(n_neighbors = k, weights = 'distance', p = 1).fit(X,y)
neigh

Now, that we've trained our model on the Manhattan neighborhoods let's use it to assign the most similar Manhattan neighborhood to each Toronto neighborhood.

In [51]:
yhat = neigh.predict(Z)
yhat

array([19, 28, 13, 26, 14, 36, 26, 39, 28, 37, 37, 30, 15,  8, 26, 30, 14,
       10, 30, 19, 39, 30, 33, 39, 16, 28,  3, 26, 21, 25, 34, 28, 31, 26,
        9,  8,  9, 19])

In [52]:
assigned_neigh = []
for n in range(yhat.shape[0]):
    assigned_neigh.append(manhattan_grouped.iloc[yhat[n]][0])

Now, let's calculate how similar a Toronto neighborhood is to the assigned Manhattan neighborhood. Looking at the Manhattan_grouped and Toronto_grouped dataframes, we can see that the values range from 0 to 1. These values represent the percentage of the total venues in each neighborhood that is of the given venue type. Thus we can see that two neighborhoods would be completely dissimilar if by subtracting their column values and adding the absolute values of these we get a value of 2. Alternatively, if doing the same process yields a value of 0, the two neighborhoods are as similar as possible within our definition of similarity. Thus, we can see that by subtracting two neighborhoods column values and adding the absolute values will obtain a value that ranges from 0 to 2. Let's subtract this value from 2 and then divide the result by 2 to obtain a percentage of how similar the neighborhoods are.

In [53]:
similarity = []
for n in range(yhat.shape[0]):
    similarity.append((2 -np.sum(np.abs(toronto_grouped_clustering.iloc[n][:]-manhattan_grouped_clustering.iloc[yhat[n]][:])))/2)

Now, we will assign the most similar Manhattan neighborhood to each Toronto neighborhood and merge this with the other Toronto data.

In [54]:
# add assigned manhattan neighborhood
toronto_sorted.insert(0, '% Similar', similarity)
toronto_sorted.insert(0, 'Manhattan Neighborhood', assigned_neigh)

In [55]:
toronto_merged = toronto

# merge toronto_grouped with toronto to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_sorted.set_index('Neighborhood'), on='Neighborhood')

Let's make sure that these choices are indeed the most similar neighborhoods as we've defined them by checking them against random choices.

In [56]:
import random

In [91]:
for n in range(yhat.shape[0]):
    print(np.sum(np.abs(toronto_grouped_clustering.iloc[n][:]-manhattan_grouped_clustering.iloc[yhat[n]][:])),np.sum(np.abs(toronto_grouped_clustering.iloc[n][:]-manhattan_grouped_clustering.iloc[random.randint(0,39)][:])))

1.2501960784313726 1.3529411764705883
1.595959595959596 1.8444444444444446
1.0921052631578947 1.4345238095238093
1.1122789028454243 1.2680722891566265
1.4999999999999998 1.5342465753424657
1.1777777777777776 1.4366013071895425
0.9205384281372123 1.2317616701853156
1.0241008527994067 1.249266862170088
1.4444444444444442 1.8181818181818181
1.5055555555555555 1.6739130434782608
1.3728070175438596 1.6629213483146068
1.6666666666666665 1.976190476190476
1.3191489361702127 1.401844532279315
1.0719640179910046 1.2739130434782608
1.0124113475177303 1.2819444444444446
1.6666666666666665 2.0
1.90625 1.9574468085106378
1.2306046013911183 1.586206896551724
1.3333333333333333 1.9166666666666667
1.2791836734693878 1.3716759431045147
1.5310344827586206 1.7857142857142856
1.3333333333333335 1.9770114942528736
1.945205479452055 2.0000000000000004
1.140886699507389 1.8285714285714285
1.0793491864831037 1.4410058027079304
1.4444444444444442 1.75609756097561
1.1034322820037106 1.2249334516415262
1.1276595

As we can see, the numbers on the right are bounded from below, i.e. they never fall less than, by the numbers on the left. Therefore, these are indeed are best choices.

In [115]:
toronto_merged

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
2                      The Beaches West, India Bazaar  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                                  North Toronto West  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                         Moore Park, Summerhill East  43.689574 -79.383160   
9   South Hill, Rathnelly, Summerhill West, Deer P...  43.686412 -79.400049   
10                                           Rosedale  43.679563 -79.377529   
11                        Cabbagetown, St. James Town  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                          Regent Park, Harbourfront  43.654260 -79.360636   
14                           Ryerson, Garden District  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                           King, Adelaide, Richmond  43.650571 -79.384568   
19  Toronto Islands, Harbourfront East, Union Station  43.640816 -79.381752   
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
23                Forest Hill North, Forest Hill West  43.696948 -79.411307   
24                The Annex, Yorkville, North Midtown  43.672710 -79.405678   
25                     University of Toronto, Harbord  43.662696 -79.400049   
26          Kensington Market, Grange Park, Chinatown  43.653206 -79.400049   
27  Railway Lands, South Niagara, Harbourfront Wes...  43.628947 -79.394420   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
29             Underground city, First Canadian Place  43.648429 -79.382280   
30                                           Christie  43.669542 -79.422564   
31                       Dovercourt Village, Dufferin  43.669005 -79.442259   
32                           T

Now, let's map the Toronto neighborhoods along with their most similar Manhattan neighborhood and % similar in parenthesis. 

In [101]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[t_latitude, t_longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood, m_neighborhood, d in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Borough'], toronto_merged['Neighborhood'], toronto_merged['Manhattan Neighborhood'], toronto_merged['% Similar']):
    label = '{}; {} ({}: {}%)'.format(neighborhood, borough, m_neighborhood,round(d,2)*100)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [136]:
toronto_merged[['Neighborhood','Manhattan Neighborhood', '% Similar']].sort_values(by = '% Similar', ascending = False)

Neighborhood Manhattan Neighborhood  \
21                     Commerce Court, Victoria Hotel            Murray Hill   
20           Toronto Dominion Centre, Design Exchange            Murray Hill   
18                           King, Adelaide, Richmond            Murray Hill   
29             Underground city, First Canadian Place     Financial District   
7                                          Davisville              Yorkville   
19  Toronto Islands, Harbourfront East, Union Station     Financial District   
26          Kensington Market, Grange Park, Chinatown           East Village   
14                           Ryerson, Garden District             Lenox Hill   
11                        Cabbagetown, St. James Town       Hamilton Heights   
15                                     St. James Town                Chelsea   
17                                 Central Bay Street            Murray Hill   
3                                     Studio District         Manhattanville   
28                    Stn A PO Boxes 25 The Esplanade            Murray Hill   
32                           Trinity, Little Portugal           East Village   
36                                 Runnymede, Swansea              Yorkville   
1                        The Danforth West, Riverdale           Sutton Place   
12                               Church and Wellesley        Upper West Side   
24                The Annex, Yorkville, North Midtown    Morningside Heights   
2                      The Beaches West, India Bazaar       Roosevelt Island   
6                                  North Toronto West               Flatiron   
16                                        Berczy Park        Lower East Side   
13                          Regent Park, Harbourfront        Lower East Side   
34                      High Park, The Junction South                 Inwood   
27  Railway Lands, South Niagara, Harbourfront Wes...        Stuyvesant Town   
10                                           Rosedale        Stuyvesant Town   
25                     University of Toronto, Harbord        Lower East Side   
33       Exhibition Place, Brockton, Parkdale Village     Washington Heights   
9   South Hill, Rathnelly, Summerhill West, Deer P...       Roosevelt Island   
5                                    Davisville North       Roosevelt Island   
30                                           Christie           Hudson Yards   
31                       Dovercourt Village, Dufferin     Washington Heights   
35                             Roncesvalles, Parkdale              Yorkville   
37  Business Reply Mail Processing Centre 969 Eastern       Roosevelt Island   
23                Forest Hill North, Forest Hill West        Stuyvesant Town   
4                                       Lawrence Park        Stuyvesant Town   
8                         Moore Park, Summerhill East           Hudson Yards   
0                                         The Beaches             Turtle Bay   
22                                           Roselawn             Tudor City   

    % Similar  
21   0.539731  
20   0.511735  
18   0.493794  
29   0.492903  
7    0.487950  
19   0.466117  
26   0.464018  
14   0.460325  
11   0.453947  
15   0.448284  
17   0.443861  
3    0.443850  
28   0.436170  
32   0.430293  
36   0.429557  
1    0.417077  
12   0.411111  
24   0.405435  
2    0.388889  
6    0.384698  
16   0.374902  
13   0.360408  
34   0.340426  
27   0.333333  
10   0.333333  
25   0.333333  
33   0.313596  
9    0.277778  
5    0.277778  
30   0.250000  
31   0.247222  
35   0.234483  
37   0.202020  
23   0.166667  
4    0.166667  
8    0.046875  
0    0.035714  
22   0.027397

## Results <a name="results"></a>

Let's now compare the Toronto neighborhood (*Commerce Court, Victoria Hotel*) with the most similarity to it's assigned Manhattan neighborhood (*Murray Hill*) to see if they share similar types and relative amounts of venues.

In [61]:
toronto_merged[toronto_merged["% Similar"] == toronto_merged["% Similar"].max()]

PostalCode           Borough                    Neighborhood   Latitude  \
21        M5L  Downtown Toronto  Commerce Court, Victoria Hotel  43.648198   

    Longitude Manhattan Neighborhood  % Similar 1st Most Common Venue  \
21 -79.379817            Murray Hill   0.539731           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                  Café                 Hotel            Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21   American Restaurant             Gastropub         Deli / Bodega   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21    Seafood Restaurant            Steakhouse     Italian Restaurant

In [62]:
manhattan_merged[manhattan_merged['Neighborhood'] == 'Murray Hill']

Borough Neighborhood   Latitude  Longitude 1st Most Common Venue  \
16  Manhattan  Murray Hill  40.748303 -73.978332           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                 Hotel        Sandwich Place   Japanese Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16    Italian Restaurant                   Gym  Gym / Fitness Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
16     French Restaurant   American Restaurant             Bagel Shop

As we can see, their most common venues are both **Coffee Shops**, **11%** of the total venues in *Commerce Court, Victoria Hotel* and **5.3%** of the total venues in *Murray Hill*. They also both seem to have a decent number of **Hotels**, **6%** of the total venues in *Commerce Court, Victoria Hotel* and **4%** of the total venues in *Murray Hill*. Within their top most common venues they also have a fair amount of **American Restaurants**, **4%** of the total venues in *Commerce Court, Victoria Hotel* and **3%** of total venues in *Murray Hill*. The final most common venue type is **Italian Restaurants** with **3%** of total venues for both *Commerce Court, Victoria Hotel* and *Murray Hill*, this is how they are most similar.

In [70]:
toronto_grouped[toronto_grouped['Neighborhood'] == 'Commerce Court, Victoria Hotel']

Neighborhood  Afghan Restaurant  American Restaurant  \
6  Commerce Court, Victoria Hotel                0.0             0.040816   

   Antique Shop  Art Gallery  Art Museum  Arts & Crafts Store  \
6           0.0     0.010204         0.0                  0.0   

   Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank  \
6          0.010204        0.0         0.0         0.0  0.020408   0.0   

        Bar  Bed & Breakfast  Beer Bar  Beer Store  Bistro  Boat or Ferry  \
6  0.020408              0.0  0.020408         0.0     0.0            0.0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Bubble Tea Shop  \
6   0.010204       0.0                   0.0        0.010204              0.0   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
6       0.0      0.020408            0.0       0.0      0.0  0.061224   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
6                        0.0           0.0                   0.0          0.0   

   Chinese Restaurant  Chocolate Shop  Climbing Gym  Clothing Store  \
6            0.010204             0.0           0.0             0.0   

   Cocktail Bar  Coffee Shop  Concert Hall  Convenience Store  Cosmetics Shop  \
6      0.010204     0.112245      0.010204                0.0             0.0   

   Coworking Space  Creperie  Cuban Restaurant  Cupcake Shop  Dance Studio  \
6              0.0  0.010204               0.0      0.010204           0.0   

   Deli / Bodega  Department Store  Dessert Shop  Dim Sum Restaurant  Diner  \
6       0.030612          0.010204           0.0                 0.0    0.0   

   Discount Store  Dog Run  Donut Shop  Dumpling Restaurant  \
6             0.0      0.0         0.0                  0.0   

   Eastern European Restaurant  Electronics Store  Ethiopian Restaurant  \
6                          0.0                0.0                   0.0   

   Event Space  Falafel Restaurant  Farmers Market  Fast Food Restaurant  \
6          0.0                 0.0             0.0              0.010204   

   Filipino Restaurant  Fish Market  Flea Market  Flower Shop  \
6                  0.0          0.0          0.0          0.0   

   Food & Drink Shop  Food Court  Food Truck  Fountain  French Restaurant  \
6                0.0    0.010204    0.010204  0.010204           0.010204   

   Fried Chicken Joint  Furniture / Home Store  Gaming Cafe  Garden  \
6             0.010204                     0.0          0.0     0.0   

   Garden Center  Gastropub  Gay Bar  General Entertainment  Gift Shop  \
6            0.0   0.030612      0.0                    0.0        0.0   

   Gourmet Shop  Greek Restaurant  Grocery Store       Gym  \
6           0.0               0.0            0.0  0.030612   

   Gym / Fitness Center  Harbor / Marina  Health & Beauty Service  \
6              0.010204              0.0                      0.0   

   Health Food Store  Historic Site  History Museum  Hobby Shop  Hookah Bar  \
6                0.0            0.0             0.0         0.0         0.0   

   Hostel     Hotel  Hotel Bar  Hotpot Restaurant  Ice Cream Shop  \
6     0.0  0.061224        0.0                0.0        0.010204   

   Indian Restaurant  Indie Movie Theater  Irish Pub  Italian Restaurant  \
6                0.0                  0.0   0.010204            0.030612   

   Japanese Restaurant  Jazz Club  Jewelry Store  Jewish Restaurant  \
6             0.020408   0.010204            0.0                0.0   

   Juice Bar  Korean Restaurant  Latin American Restaurant  Lingerie Store  \
6        0.0                0.0                   0.010204             0.0   

   Liquor Store  Lounge  Malay Restaurant  Market  Martial Arts Dojo  \
6           0.0     0.0               0.0     0.0                0.0   

   Mediterranean Restaurant  Men's Store  Mexican Restaurant  \
6                       0.0          0.0                 0.0   

   Middle Eastern Restaurant  Miscella

In [71]:
manhattan_grouped[manhattan_grouped['Neighborhood'] == 'Murray Hill']

Neighborhood  Afghan Restaurant  American Restaurant  Antique Shop  \
26  Murray Hill                0.0             0.031915           0.0   

    Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  BBQ Joint  \
26          0.0         0.0                  0.0               0.0        0.0   

    Baby Store  Bagel Shop    Bakery  Bank       Bar  Bed & Breakfast  \
26         0.0    0.021277  0.021277   0.0  0.021277              0.0   

    Beer Bar  Beer Store  Bistro  Boat or Ferry  Bookstore  Boutique  \
26       0.0         0.0     0.0            0.0        0.0       0.0   

    Brazilian Restaurant  Breakfast Spot  Bubble Tea Shop  Building  \
26                   0.0             0.0              0.0  0.010638   

    Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
26      0.021277            0.0       0.0      0.0  0.010638   

    Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  \
26                        0.0           0.0                   0.0   

    Cheese Shop  Chinese Restaurant  Chocolate Shop  Climbing Gym  \
26          0.0            0.021277             0.0           0.0   

    Clothing Store  Cocktail Bar  Coffee Shop  Concert Hall  \
26             0.0      0.021277     0.053191           0.0   

    Convenience Store  Cosmetics Shop  Coworking Space  Creperie  \
26                0.0             0.0              0.0       0.0   

    Cuban Restaurant  Cupcake Shop  Dance Studio  Deli / Bodega  \
26          0.021277      0.010638           0.0       0.010638   

    Department Store  Dessert Shop  Dim Sum Restaurant     Diner  \
26               0.0           0.0                 0.0  0.021277   

    Discount Store  Dog Run  Donut Shop  Dumpling Restaurant  \
26             0.0      0.0         0.0                  0.0   

    Eastern European Restaurant  Electronics Store  Ethiopian Restaurant  \
26                          0.0                0.0                   0.0   

    Event Space  Falafel Restaurant  Farmers Market  Fast Food Restaurant  \
26     0.010638                 0.0             0.0                   0.0   

    Filipino Restaurant  Fish Market  Flea Market  Flower Shop  \
26                  0.0          0.0          0.0          0.0   

    Food & Drink Shop  Food Court  Food Truck  Fountain  French Restaurant  \
26                0.0         0.0    0.010638       0.0           0.031915   

    Fried Chicken Joint  Furniture / Home Store  Gaming Cafe  Garden  \
26             0.010638                0.010638          0.0     0.0   

    Garden Center  Gastropub  Gay Bar  General Entertainment  Gift Shop  \
26            0.0   0.010638      0.0                    0.0        0.0   

    Gourmet Shop  Greek Restaurant  Grocery Store       Gym  \
26           0.0               0.0       0.010638  0.031915   

    Gym / Fitness Center  Harbor / Marina  Health & Beauty Service  \
26              0.031915              0.0                 0.010638   

    Health Food Store  Historic Site  History Museum  Hobby Shop  Hookah Bar  \
26                0.0            0.0             0.0         0.0         0.0   

    Hostel     Hotel  Hotel Bar  Hotpot Restaurant  Ice Cream Shop  \
26     0.0  0.042553        0.0                0.0             0.0   

    Indian Restaurant  Indie Movie Theater  Irish Pub  Italian Restaurant  \
26                0.0                  0.0        0.0            0.031915   

    Japanese Restaurant  Jazz Club  Jewelry Store  Jewish Restaurant  \
26             0.042553   0.010638            0.0           0.021277   

    Juice Bar  Korean Restaurant  Latin American Restaurant  Lingerie Store  \
26   0.010638                0.0                   0.010638             0.0   

    Liquor Store    Lounge  Malay Restaurant  Market  Martial Arts Dojo  \
26      0.010638  0.010638               0.0     0.0           0.010638   

    Mediterranean Restaurant  Men's Store  Mexican Restaurant  \
26                  0.021277          0.0      

Now, let's look at the Toronto neighborhood (*Roselawn*) with the least similarity to it's assigned Manhattan neighborhood (*Tudor City*).

In [84]:
toronto_merged[toronto_merged["% Similar"] == toronto_merged["% Similar"].min()]

PostalCode          Borough Neighborhood   Latitude  Longitude  \
22        M5N  Central Toronto     Roselawn  43.711695 -79.416936   

   Manhattan Neighborhood  % Similar 1st Most Common Venue  \
22             Tudor City   0.027397                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Yoga Studio  Ethiopian Restaurant           Flower Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22           Flea Market           Fish Market   Filipino Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22  Fast Food Restaurant        Farmers Market     Falafel Restaurant

In [85]:
manhattan_merged[manhattan_merged['Neighborhood'] == 'Tudor City']


Borough Neighborhood   Latitude  Longitude 1st Most Common Venue  \
36  Manhattan   Tudor City  40.746917 -73.971219                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
36                  Café    Mexican Restaurant      Greek Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
36      Sushi Restaurant         Deli / Bodega           Pizza Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
36       Thai Restaurant                 Diner       Asian Restaurant

As we can see, none of the top most common venue types are shared between the two. However, this is the most similar Manhattan neighborhood by our definition. Therefore, we can conclude that this neighborhood is quite different from all Manhattan neighborhoods.

Now, let's calculate the overall similarity between Toronto proper and Manhattan by averaging the % similarity of every Toronto neighborhood with it's assigned Manhattan neighborhood.

In [99]:
np.mean(toronto_merged['% Similar'])

0.35083361859623813

In [106]:
np.mean(toronto_merged[toronto_merged['Borough'] == 'Downtown Toronto']['% Similar'])

0.42262816166614176

In [107]:
np.mean(toronto_merged[toronto_merged['Borough'] == 'Central Toronto']['% Similar'])

0.2490270227418106

In [108]:
np.mean(toronto_merged[toronto_merged['Borough'] == 'West Toronto']['% Similar'])

0.33259609538154694

In [109]:
np.mean(toronto_merged[toronto_merged['Borough'] == 'East Toronto']['% Similar'])

0.2975101639401842

With an overall similarity of about **35%** we can assume that in fact Toronto is not that similar to Manhattan overall. However, if we focus on each borough of Toronto we find that *Downtown Toronto* is overall **42%** similar. While the other boroughs are overall less similar than the total average with *Central Toronto* at **25%**, *West Toronto* at **33%** and *East Toronto* at **30%**.

## Discussion <a name="discussion"></a>

Through our analysis using Foursquare API data and by only looking at venue types that are common between Manhattan and Toronto proper, it can be concluded that Toronto is not very similar to Manhattan. In particular, Manhattan neighborhoods seem to be defined by the type of food available in them while Toronto neighborhoods seem to be far less so.
To answer the original question, can we recommend a neighborhood of Manhattan that is similar to each neighborhood of Toronto, the answer appears to be no. However, we can make some fairly good recommendations for neighborhoods for Downtown Toronto in particular.
However, there are a few ways we could improve our analysis and make a better recommendation.
We could lump categories of food, drinking establishments and the like into one category.
We could narrow the categories we look at, for example only retail establishments or only eateries based on the stakeholders needs.


## Conclusion <a name="conclusion"></a>

The purpose of this project was to find neighborhoods in Manhattan which are similar to neighborhoods in Toronto Proper. In order to give recommendations to people and companies moving from Toronto to Manhattan and looking for a similar neighborhood. We have come to the conclusion that the neighborhoods of these two cities are in fact not very similar overall but we can make the best recommendation for each neighborhood with the data provided by Foursquare.
As an extension to this project, it would be a good exercise to look at other large cities and see how well they compare to Manhattan and Toronto. Perhaps London or Paris would make a more suitable match to Toronto.